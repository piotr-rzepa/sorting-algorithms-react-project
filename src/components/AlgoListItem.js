/*
 * Komponent bezklasowy reprezentujący pojedynczy algorytm w liście
 */

/* eslint-disable react/prop-types */
import React from 'react';
import { Link } from 'react-router-dom';

export default function AlgoListItem(props) {
	return (
		<React.Fragment key={props.id}>
			<li>
				<Link to={`/algorithm/${props.name}`}>{props.name}</Link>:{' '}
				{props.sortingTime.average}{' '}
			</li>
		</React.Fragment>
	);
}
