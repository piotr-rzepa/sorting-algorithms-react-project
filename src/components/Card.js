/*
 * Komponent bezklasowy będący "kartą" -> Blokiem z informacją np. o  szybkości działania
 */

/* eslint-disable react/prop-types */
import React from 'react';

export default function Card({ title, subtitle }) {
	return (
		<div className="Card">
			<h2>{title}</h2>
			<p>{subtitle}</p>
		</div>
	);
}
