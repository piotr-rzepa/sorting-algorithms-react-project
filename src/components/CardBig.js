/*
 * Komponent bezklasowy -> wariacja karty, z innym stylem
 */

/* eslint-disable react/prop-types */
import React from 'react';

export default function CardBig({ desc }) {
	return (
		<div className="Card Card--big">
			<h4>{desc}</h4>
		</div>
	);
}
