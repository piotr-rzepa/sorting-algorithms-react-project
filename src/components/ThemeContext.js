/* eslint-disable no-unused-vars */
/* eslint-disable react/prop-types */
import React, { useState, useContext, createContext } from 'react';

//Tworzymy kontekst do przekazywania informacji o trybie jasnym/ciemnym
export const ThemeContext = createContext();
//Tworzymy kontekst do ustawiania trybu -> pod przyciskiem na pasku wertykalnym
export const ThemeContextUpdate = createContext();

//Komponent który będzie zapewniał wszystkim {children} dostęp do informacji o motywie
export const ThemeProvider = ({ children }) => {
	const [darkTheme, setDarkTheme] = useState(false);

	//Metoda do zmieany stylu
	const updateContext = () => setDarkTheme((prevTheme) => !prevTheme);

	return (
		<ThemeContext.Provider value={darkTheme}>
			<ThemeContextUpdate.Provider value={updateContext}>
				{children}
			</ThemeContextUpdate.Provider>
		</ThemeContext.Provider>
	);
};
