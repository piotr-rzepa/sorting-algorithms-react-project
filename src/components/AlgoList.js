/*
 * Komponent bezklasowy odpowiedzialny za listę algorytmów w Dashboardzie
 */

/* eslint-disable react/prop-types */
import React from 'react';
import AlgoListItem from './AlgoListItem';
import { useSelector, useDispatch } from 'react-redux';
import selectSortedAlgorithms from '../selectors/sortingSelector';
import { setSortingAsc, setSortingDesc } from '../actions/filters';

export default function AlgoList() {
	const algorithms = useSelector((state) => selectSortedAlgorithms(state));
	const dispatch = useDispatch();

	//* Zmiana sortowania z dispatchem do reduxa
	const onSortChange = (sort) => {
		sort === 'Descending'
			? dispatch(setSortingDesc())
			: dispatch(setSortingAsc());
	};

	return (
		<div className="algorithm-list">
			<div className="algorithm-list-title">
				<p>Algorithms by sorting time</p>
				<select
					className="select"
					name="sorting"
					onChange={(e) => onSortChange(e.target.value)}
				>
					<option>Descending</option>
					<option>Ascending</option>
				</select>
			</div>
			<hr />
			<ul>
				{algorithms.map((algorithm) => (
					<AlgoListItem
						key={algorithm.id}
						name={algorithm.name}
						sortingTime={algorithm.sortingTime}
					/>
				))}
			</ul>
		</div>
	);
}
