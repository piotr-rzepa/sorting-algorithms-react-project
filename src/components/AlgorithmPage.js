/*
 * Komponent bezklasowy będący stroną z informacjami o poszczególnym algorytmie
 */

/* eslint-disable no-unused-vars */
/* eslint-disable react/prop-types */
import React, { useState, useEffect, useContext } from 'react';
import { useSelector } from 'react-redux';
import Card from './Card';
import CardBig from './CardBig';
import ChartLine from './ChartLine';
import ChartVisualize from './ChartVisualize';
import SettingsModal from './SettingsModal';
import { ReactComponent as PlusSVG } from '../images/plus.svg';
import { ReactComponent as EditSVG } from '../images/edit.svg';
import InfoSVG from '../images/info.svg';
import PercentSVG from '../images/percent.svg';
import RefreshSVG from '../images/refresh.svg';
import ClockSVG from '../images/clock.svg';
import TrophySVG from '../images/trophy.svg';
import IncreaseSVG from '../images/increase.svg';
import { evaluateResults } from '../sortingImpl/sortManager';
import { useDispatch } from 'react-redux';
import { setSortingTime } from '../actions/sorting';
import numeral from 'numeral';
import worker from 'workerize-loader?inline!../workers/worker';
import { selectSortedPositions } from '../selectors/sortingSelector';
import { ThemeContext } from './ThemeContext';
import { styleMain } from '../styles/styles';
//* Obiekt match zawiera informację o ścieżce, parametrach, (w tym nazwę algorytmu który chcemy wyświetlić)
export default function AlgorithmPage({ match }) {
	// Pobieranie motywu [jasny/ciemny]
	const darkTheme = useContext(ThemeContext);

	//Ustawianie motywu w zależności od kontekstu
	const currentStyle = darkTheme ? styleMain.darkStyle : styleMain.lightStyle;
	//* State naszego workera
	const [instance, setInstance] = useState(undefined);
	const [canRun, setCanRun] = useState({
		computation: true,
		visualization: true
	});

	//* Tylko przy montowaniu komponentu tworzymy workera, przy unmount usuwamy go
	useEffect(() => {
		setInstance(new worker());
		return () => setInstance(undefined);
	}, []);

	//* Tworzenie losowych danych do wizualizacji
	let test = Array.from({ length: 150 }, () => Math.floor(Math.random() * 100));

	//* Obiekt algorytmu, na którego stronie jesteśmy
	const sortObj = useSelector((state) => state.algorithms).filter(
		(sort) => sort.name === match.params.id
	)[0];
	const place = useSelector((state) => selectSortedPositions(state)).indexOf(
		sortObj
	);

	const dispatch = useDispatch();

	const startingData = {
		sortingPercent: 0,
		sortingStep: 0,
		data: test.map((value) => ({ pv: value }))
	};
	/*
	 * data, setDate -> manipulowanie danymi wizualizacji
	 * lineChartData, setLineChartData -> Dane do wykresu liniowego
	 * sortingInfo, setSortingInfo -> Dane podczas iteracji sortowania
	 * modalIsOpen, setIsOpen -> wyświetlanie okna ustawień
	 */
	const [data, setData] = useState(startingData);
	const [lineChartData, setLineChartdata] = useState([]);
	const [sortingInfo, setSortingInfo] = useState({ passed: 0, time: 0 });
	const [modalIsOpen, setIsOpen] = useState(false);

	const options = useSelector((state) => state.tests);
	const sortT = sortObj.sortingTime.average;

	//* Metoda odpowiedzialna za wizualizacje sortowania
	const visualizationStep = () => {
		setCanRun({ ...canRun, visualization: false });
		const arr = [];

		//* Pobieramy stan sortowanej tablicy po każdej iteracji algorytmu sortującego
		let gen = sortObj.generator(test);
		for (let vStep of gen) {
			arr.push(Array.from(vStep));
		}

		//* Wyświetlamy krok wizualizacji oraz % ukończenia w  komponentach Card
		arr.forEach((ele, i) => {
			setTimeout(() => {
				setData({
					sortingStep: i + 1,
					sortingPercent: (i + 1) / arr.length,
					data: ele.map((value) => ({ pv: value }))
				});
				//* Po skończeniu wizualizacji odblokowujemy przycisk
				if (i === arr.length - 1) {
					setCanRun({ ...canRun, visualization: true });
				}
			}, 300 * i);
		});
	};

	//* Metoda odpowiedzialna za przeprowadzenie testów sortowania
	const runTestRoutine = async () => {
		setCanRun({ ...canRun, computation: false });
		let passed = 0;
		const lineData = [];

		//* Uruchamiamy sortowanie w osobnym procesie przez workera
		const arr = await instance.sortManager(match.params.id, options);

		//* Wyświetlamy krok oraz czas każdego ze stanów w komponentach Card
		arr.forEach((ele, i) => {
			setTimeout(() => {
				setSortingInfo({
					passed: ++passed,
					time: ele.average
				});
				//* Po skończeniu wizualizacji odblokowujemy przycisk
				if (i === arr.length - 1) {
					setCanRun({ ...canRun, computation: true });
				}
			}, 200 * i);
			lineData.push({ ...ele });
		});

		//* Wyświetlamy dane na wykresie
		setTimeout(() => {
			const [best, average, worst] = evaluateResults(lineData);
			setLineChartdata(lineData);
			dispatch(setSortingTime(match.params.id, { worst, best, average }));
		}, arr.length * 200);
	};

	return (
		<div className="Dashboard" style={currentStyle}>
			<SettingsModal
				isOpen={modalIsOpen}
				onRequestClose={() => setIsOpen(false)}
				contentLabel="Example Modal"
			/>
			<div className="overview">
				<div className="overview-title">
					<h1>{match.params.id}</h1>
					<p>Overview of sorting algorithm</p>
				</div>
				<div className="button-group">
					<button
						type="button"
						className="data-options-button"
						onClick={() => setIsOpen(() => true)}
					>
						<EditSVG width="24px" height="24px" />
						Manage data options
					</button>
					<button
						className={`data-options-button ${
							!canRun.computation ? 'button--disabled' : ''
						}`}
						type="button"
						onClick={() => runTestRoutine()}
					>
						<PlusSVG width="24px" height="24px" />
						Run tests
					</button>
					<button
						className={`data-options-button ${
							!canRun.visualization ? 'button--disabled' : ''
						}`}
						type="button"
						onClick={() => visualizationStep()}
					>
						<PlusSVG width="24px" height="24px" />
						Show example visualization
					</button>
				</div>
			</div>
			<div className="small-cards algorithms">
				<Card
					title={`#${place + 1}`}
					subtitle={'Position in ranking'}
					icon={TrophySVG}
					background={{ backgroundColor: 'rgba(255,184,0,0.07)' }}
				/>
				<Card
					title={`${sortT}ms`}
					subtitle={'Average sorting time'}
					icon={ClockSVG}
					background={{ backgroundColor: 'rgba(76,184,255,0.07)' }}
				/>
				<CardBig
					desc={sortObj.description}
					icon={InfoSVG}
					background={{ backgroundColor: 'white' }}
				/>
			</div>
			<div className="charts">
				<ChartLine width={820} height={380} data={lineChartData} />
				<ChartVisualize width={890} height={404} data={data.data} />
			</div>
			<div className="small-cards">
				<Card
					title={sortingInfo.passed}
					subtitle={'Sortings passed'}
					icon={RefreshSVG}
					background={{ backgroundColor: 'rgba(255,184,0,0.07)' }}
				/>
				<Card
					title={sortingInfo.time}
					subtitle={'Sorting time'}
					icon={ClockSVG}
					background={{ backgroundColor: 'rgba(76,184,255,0.07)' }}
				/>
				<Card
					title={data.sortingStep}
					subtitle={'Sorting step'}
					icon={IncreaseSVG}
					background={{ backgroundColor: 'rgba(76,184,255,0.07)' }}
				/>
				<Card
					title={numeral(data.sortingPercent).format('0.00%')}
					subtitle={'Sorting completion'}
					icon={PercentSVG}
					background={{ backgroundColor: 'rgba(76,184,255,0.07)' }}
				/>
			</div>
			<p style={{ textAlign: 'right' }}>Made by Chili Labs & Piotr</p>
		</div>
	);
}
