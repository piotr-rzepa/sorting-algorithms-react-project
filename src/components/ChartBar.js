/* eslint-disable no-unused-vars */
/*
 * Komponent bezklasowy będacy wykresem słupkowym z czasami worst,best,average \
 * poszczególnych algorytmów na stronie głównej
 */

/* eslint-disable react/prop-types */
import React, { useContext } from 'react';
import {
	BarChart,
	Bar,
	XAxis,
	YAxis,
	Tooltip,
	Legend,
	CartesianGrid,
	LabelList,
	ResponsiveContainer
} from 'recharts';
import { useSelector } from 'react-redux';
import { ThemeContext } from './ThemeContext';
import { styleCard } from '../styles/styles';
export default function ChartBar(props) {
	// Pobieranie motywu [jasny/ciemny]
	const darkTheme = useContext(ThemeContext);

	//Ustawianie motywu w zależności od kontekstu
	const currentStyle = darkTheme ? styleCard.darkStyle : styleCard.lightStyle;

	const barColors = ['#6F52ED', '#28E451', '#FF7A00'];
	const algorithms = useSelector((state) => state.algorithms);

	const chartData = [...algorithms].map((algorithm) => {
		return {
			name: algorithm.name,
			best: algorithm.sortingTime.best,
			worst: algorithm.sortingTime.worst,
			average: algorithm.sortingTime.average
		};
	});
	return (
		<div
			className="card-chart card-chart--big"
			style={{ height: props.height, ...currentStyle }}
		>
			<ResponsiveContainer width="100%" height="100%">
				<BarChart width={props.width} height={props.height} data={chartData}>
					<CartesianGrid strokeDasharray="3 3" />
					<XAxis dataKey="name" />
					<YAxis />
					<Tooltip />
					<Legend align="right" verticalAlign="top" />
					{Object.getOwnPropertyNames(algorithms[0].sortingTime).map(
						(property, index) => (
							<Bar dataKey={property} key={property} fill={barColors[index]}>
								<LabelList dataKey={property} position="top" />
							</Bar>
						)
					)}
				</BarChart>
			</ResponsiveContainer>
		</div>
	);
}
