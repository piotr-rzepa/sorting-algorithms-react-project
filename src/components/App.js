/*
 * Komponent bezklasowy będący rootem naszej aplikacji
 */
// @flow
import React from 'react';
import HorizontalNav from './HorizontalNav';
import VerticalNav from './VerticalNav';
import Dashboard from './Dashboard';
import AlgorithmPage from './AlgorithmPage';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { ThemeProvider } from './ThemeContext';
const App = () => {
	return (
		<ThemeProvider>
			<Router>
				<div>
					<HorizontalNav />
					<VerticalNav />
					<Switch>
						<Route path="/" exact component={Dashboard} />
						<Route path="/algorithm/:id" exact component={AlgorithmPage} />
					</Switch>
				</div>
			</Router>
		</ThemeProvider>
	);
};

export default App;
