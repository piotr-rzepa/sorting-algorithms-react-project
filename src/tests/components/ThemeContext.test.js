import * as React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from '../../components/ThemeContext';

describe('component handles theme change', () => {
	const setState = jest.fn();
	const useStateMock = (initState) => [initState, setState];
	const createContext = jest.spyOn(React, 'createContext');
	afterEach(() => {
		jest.clearAllMocks();
		createContext.mockClear();
	});
	it('Renders correctly', () => {
		const darkTheme = jest.fn();
		jest.spyOn(React, 'useState').mockImplementation(useStateMock);
		const wrapper = shallow(<ThemeProvider value={darkTheme} />);
		expect(wrapper).toMatchSnapshot();
	});
});
