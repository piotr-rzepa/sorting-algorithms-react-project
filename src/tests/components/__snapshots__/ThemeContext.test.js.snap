// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component handles theme change Renders correctly 1`] = `
<ContextProvider
  value={false}
>
  <ContextProvider
    value={[Function]}
  />
</ContextProvider>
`;
